components/brave_wallet/browser/asset_discovery_manager.cc

```diff
@@ -405,13 +335,125 @@ void AssetDiscoveryManager::MergeDiscoveredEthAssets(
+    }
+  }
+
+  if (solana_addresses.empty()) {
+    std::move(callback).Run({});
+    return;
+  }
+
+  const auto barrier_callback =
+      base::BarrierCallback<std::vector<SolanaAddress>>(
+          solana_addresses.size(),
+          base::BindOnce(&AssetDiscoveryTask::MergeDiscoveredSPLTokens,
+                         weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+  for (const auto& account_address : solana_addresses) {
+    json_rpc_service_->GetSolanaTokenAccountsByOwner(
+        account_address,
+        base::BindOnce(&AssetDiscoveryTask::OnGetSolanaTokenAccountsByOwner,
+                       weak_ptr_factory_.GetWeakPtr(),
+                       std::move(barrier_callback)));
+  }
+}
+
+void AssetDiscoveryTask::OnGetSolanaTokenAccountsByOwner(
+    base::OnceCallback<void(std::vector<SolanaAddress>)> barrier_callback,
+    const std::vector<SolanaAccountInfo>& token_accounts,
+    mojom::SolanaProviderError error,
+    const std::string& error_message) {
+  if (error != mojom::SolanaProviderError::kSuccess || token_accounts.empty()) {
+    std::move(barrier_callback).Run(std::vector<SolanaAddress>());
+    return;
+  }
+
+  // Add each token account to the all_discovered_contract_addresses list
+  std::vector<SolanaAddress> discovered_mint_addresses;
+  for (const auto& token_account : token_accounts) {
+    // Decode Base64
+    const absl::optional<std::vector<uint8_t>> data =
+        base::Base64Decode(token_account.data);
+    if (data.has_value()) {
+      // Decode the address
+      const absl::optional<SolanaAddress> mint_address =
+          DecodeMintAddress(data.value());
+      if (mint_address.has_value()) {
+        // Add the contract address to the list
+        discovered_mint_addresses.push_back(mint_address.value());
+      }
+    }
+  }
+
+  std::move(barrier_callback).Run(discovered_mint_addresses);
+}
+
+void AssetDiscoveryTask::MergeDiscoveredSPLTokens(
+    DiscoverAssetsCompletedCallback callback,
+    const std::vector<std::vector<SolanaAddress>>&
+        all_discovered_contract_addresses) {
+  // Create vector of all discovered contract addresses
+  std::vector<std::string> discovered_mint_addresses;
+  for (const auto& discovered_contract_address_list :
+       all_discovered_contract_addresses) {
+    for (const auto& discovered_contract_address :
+         discovered_contract_address_list) {
+      discovered_mint_addresses.push_back(
+          discovered_contract_address.ToBase58());
+    }
+  }
+
+  // Convert vector to flat_set
+  base::flat_set<std::string> discovered_mint_addresses_set(
+      std::move(discovered_mint_addresses));
+
+  auto internal_callback =
+      base::BindOnce(&AssetDiscoveryTask::OnGetSolanaTokenRegistry,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback),
+                     std::move(discovered_mint_addresses_set));
+
+  // Fetch SOL registry tokens
+  BlockchainRegistry::GetInstance()->GetAllTokens(mojom::kSolanaMainnet,
```

Comment:
Please add a comment for this being mainnet only and needs to be changed when we are going to support multiple chains for Solana.